data_model/src/isi.rs

```diff
@@ -79,790 +65,829 @@ pub mod model {
+    pub fn asset(asset_id: AssetId) -> Self {
+        Self::Asset(Unregister::new(asset_id))
+    }
+
+    /// Constructs a new [`UnregisterBox`] for a [`Role`].
+    pub fn role(role_id: RoleId) -> Self {
+        Self::Role(Unregister::new(role_id))
+    }
+
+    /// Constructs a new [`UnregisterBox`] for a [`Trigger`].
+    pub fn trigger(trigger_id: TriggerId) -> Self {
+        Self::Trigger(Unregister::new(trigger_id))
     }
 }
 
-impl RevokeExpr {
-    /// Generic constructor.
-    pub fn new<P: Into<EvaluatesTo<Value>>, I: Into<EvaluatesTo<AccountId>>>(
-        object: P,
-        destination_id: I,
-    ) -> Self {
-        Self {
-            destination_id: destination_id.into(),
-            object: object.into(),
-        }
+isi_box! {
+    /// Enum with all supported [`Mint`] instructions.
+    pub enum MintBox {
+        /// Mint for [`Account`].
+        Account(AccountMintBox),
+        /// Mint for [`Asset`].
+        Asset(AssetMintBox),
+        /// Mint [`Trigger`] repetitions.
+        TriggerRepetitions(Mint<u32, Trigger<TriggeringFilterBox>>),
     }
 }
 
-impl GrantExpr {
-    /// Constructor.
-    pub fn new<P: Into<EvaluatesTo<Value>>, I: Into<EvaluatesTo<AccountId>>>(
-        object: P,
-        destination_id: I,
-    ) -> Self {
-        Self {
-            destination_id: destination_id.into(),
-            object: object.into(),
-        }
+isi_box! {
+    /// Enum with all supported [`Mint`] instructions related to [`Account`].
+    pub enum AccountMintBox {
+        /// Mint [`PublicKey`].
+        PublicKey(Mint<PublicKey, Account>),
+        /// Mint [`SignatureCheckCondition`].
+        SignatureCheckCondition(Mint<SignatureCheckCondition, Account>),
     }
 }
 
-impl SetKeyValueExpr {
-    /// Construct [`SetKeyValueExpr`].
-    pub fn new<
-        I: Into<EvaluatesTo<IdBox>>,
-        K: Into<EvaluatesTo<Name>>,
-        V: Into<EvaluatesTo<Value>>,
-    >(
-        object_id: I,
-        key: K,
-        value: V,
-    ) -> Self {
-        Self {
-            object_id: object_id.into(),
-            key: key.into(),
-            value: value.into(),
-        }
+isi_box! {
+    /// Enum with all supported [`Mint`] instructions related to [`Asset`].
+    pub enum AssetMintBox {
+        /// Mint [`Asset`] of [`Quantity`] type.
+        Quantity(Mint<u32, Asset>),
+        /// Mint [`Asset`] of [`BigQuantity`] type.
+        BigQuantity(Mint<u128, Asset>),
+        /// Mint [`Asset`] of [`Fixed`] type.
+        Fixed(Mint<Fixed, Asset>),
     }
 }
 
-impl RemoveKeyValueExpr {
-    /// Construct [`RemoveKeyValueExpr`].
-    pub fn new<I: Into<EvaluatesTo<IdBox>>, K: Into<EvaluatesTo<Name>>>(
-        object_id: I,
-        key: K,
+impl MintBox {
+    /// Constructs a new [`MintBox`] to mint [`PublicKey`] for [`Account`].
+    pub fn account_public_key(public_key: PublicKey, account_id: AccountId) -> Self {
+        Self::Account(AccountMintBox::PublicKey(Mint::new(public_key, account_id)))
+    }
```

Comment:
I've been looking at the API and have been very unhappy to see that users have to deal with `MintBox` or any other `*Box`. Let me present an idea how we can simplify it:

```suggestion
impl Mint {
    /// Constructs a new [`MintBox`] to mint [`PublicKey`] for [`Account`].
    pub fn account_public_key(public_key: PublicKey, account_id: AccountId) -> MintBox {
        MintBox::Account(AccountMintBox::PublicKey(Mint::new(public_key, account_id)))
    }
```

Can we define constructors on the `Mint` struct and have it return `MintBox`?

Reply:
do we even need `MintBox::account_public_key`. Can't we just have `Mint::new` and an `impl `Into<InstructionBox> for Mint`?

Reply:
`*Box` aren't just for serialization, they represent the entrypoint to our instructions

Reply:
I don't very like the approach of constructing different types in another type constructor. I'd like to open a type documentation and see all possible creation methods

Reply:
> *Box aren't just for serialization, they represent the entrypoint to our instructions

what do you mean by entrypoint for instructions?

> I don't very like the approach of constructing different types in another type constructor. I'd like to open a type documentation and see all possible creation methods

I can agree with that. That's why I suggested to implement `From<Mint> for InstructionBox`

Reply:
I tried `Mint::new()` approach and it doesn't work well because of our generic usage style (type has generic `T`, but takes `T::Id` in `new()`). It forces user to **always** specify generics like `Mint::<u32, Asset>::new(...)`.

So I decided to go with `Mint::asset_quantity()` approach (note that it's not `MintBox`). It has the following advantages:

1. Doesn't require expicit generics
2. We can identify the exact instruction type by reading function name
3. Easy to search in the project for exact instructions usages

`From` approaches works normally. So now users are disallowed to use types like `MintBox` as they serve only for (de-)serialization